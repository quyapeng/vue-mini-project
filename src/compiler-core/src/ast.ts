export enum NodeTypes {
  INTERPOLATION,
  SIMPLE_EXPRESSION,
  ELEMENT,
  TEXT,
  ROOT,
  COMPOUND_EXPRESSION,
}
