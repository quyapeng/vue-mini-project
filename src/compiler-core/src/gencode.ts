export function generate(ast) {
  const context = createCodegenContext();
  const { push } = context;
  push("return ");

  const FunctionName = "render";
  const args = ["_ctx", "_cache"];
  const signature = args.join(", ");
  push(`function ${FunctionName}(${signature}){`);
  push("return ");
  genNode(ast.codegenNode, context);
  push("}");
  //   code += `function ${FunctionName}(${signature}){`;
  //   // 介于灵活考虑，这一部分放在transform中处理好之后再返回，此处直接使用。
  //   //   console.log("ast", ast);
  //   //   const node = ast.children[0];
  //   const node = ast.codegenNode;
  //   code += `return '${node.content}'`;
  //   code += "}";

  return {
    code: context.code,
  };
}
function genNode(node: any, context) {
  // const node = ast.codegenNode;
  //   code += `return '${node.content}'`;
  const { push } = context;
  push(`'${node.content}'`);
}
function createCodegenContext() {
  const context = {
    code: "",
    push(source) {
      context.code += source;
    },
  };
  return context;
}
