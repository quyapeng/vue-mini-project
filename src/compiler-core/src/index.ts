// 统一管理 出口导出
export * from "./compiler";
