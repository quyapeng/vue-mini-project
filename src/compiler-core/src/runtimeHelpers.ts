export const TO_DISPLAY_STRING = "toDisplayString";

export const helperMapName = {
  [TO_DISPLAY_STRING]: "toDisplayString",
};
