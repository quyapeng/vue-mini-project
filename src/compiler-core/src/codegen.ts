import { isString } from "../../shared";
import { NodeTypes } from "./ast";
import {
  CREATE_ELEMENT_VNODE,
  helperMapName,
  TO_DISPLAY_STRING,
} from "./runtimeHelpers";

// 模块职责划分清晰明确
export function generate(ast) {
  const context = createCodegenContext();
  const { push } = context;
  // 导入逻辑抽取
  genFunctionPreamble(ast, context);
  //   const VueBinging = "Vue";
  //   //   const helpers = ["toDisplayString"];
  //   //   const aliasHelpers = (s: any) => `${s}: _${s}`;
  //   const helpers = ["toDisplayString"];
  //   const aliasHelpers = (s: any) => `${s}: _${s}`;
  //   // 插值 render 导入
  //   //   push(`const { ${helpers.map(aliasHelpers)} } = ${VueBinging};`);
  //   push(`const { ${ast.helpers.map(aliasHelpers)} } = ${VueBinging};`);
  //   push("\n");
  //   push("return ");

  const FunctionName = "render";
  const args = ["_ctx", "_cache"];
  const signature = args.join(", ");
  // console.log("ast", ast);

  push(`function ${FunctionName}(${signature}){`);
  push("return ");
  genNode(ast.codegenNode, context);
  push("}");
  //   code += `function ${FunctionName}(${signature}){`;
  //   // 介于灵活考虑，这一部分放在transform中处理好之后再返回，此处直接使用。
  //   //   console.log("ast", ast);
  //   //   const node = ast.children[0];
  //   const node = ast.codegenNode;
  //   code += `return '${node.content}'`;
  //   code += "}";

  return {
    code: context.code,
  };
}
function genFunctionPreamble(ast: any, context: any) {
  const { push } = context;
  const VueBinging = "Vue";
  //   const helpers = ["toDisplayString"];
  //   const aliasHelpers = (s: any) => `${s}: _${s}`;
  const aliasHelpers = (s: any) => `${helperMapName[s]}: _${helperMapName[s]}`;
  // 插值 render 导入
  //   push(`const { ${helpers.map(aliasHelpers)} } = ${VueBinging};`);
  if (ast.helpers.length > 0)
    push(
      `const { ${ast.helpers.map(aliasHelpers).join(", ")} } = ${VueBinging};`
    );
  push("\n");
  push("return ");
}
function createCodegenContext() {
  const context = {
    code: "",
    push(source) {
      context.code += source;
    },
    helper(key) {
      return `_${helperMapName[key]}`;
    },
  };
  return context;
}

function genNode(node: any, context) {
  switch (node.type) {
    case NodeTypes.TEXT:
      genText(node, context);
      break;
    case NodeTypes.INTERPOLATION:
      genInterpolation(node, context);
      break;
    case NodeTypes.SIMPLE_EXPRESSION:
      genExpression(node, context);
      break;
    case NodeTypes.ELEMENT:
      genElement(node, context);
      break;
    case NodeTypes.COMPOUND_EXPRESSION:
      genCompoundExpression(node, context);

    default:
      break;
  }
}

function genCompoundExpression(node: any, context: any) {
  const { push } = context;
  const { children } = node;
  for (let i = 0; i < children.length; i++) {
    const child = children[i];
    if (isString(child)) {
      push(child);
    } else {
      genNode(child, context);
    }
  }
}
function genElement({ tag, children, props }: any, context: any) {
  //
  const { push, helper } = context;
  // console.log("children", children);
  push(`${helper(CREATE_ELEMENT_VNODE)}("${tag}"), ${props},`);
  // console.log("llll", genNullable([tag, props, children]));
  genNodeList(genNullable([tag, props, children]), context);
  // const child = children[0];
  // genNode(children, context);
  // for (let i = 0; i < children.length; i++) {
  //   const child = children[i];
  //   genNode(child, context);
  // }
  push(")");
}

function genNodeList(nodes: any, context) {
  // console.log("nodes", nodes);
  const { push } = context;
  for (let i = 0; i < nodes.length; i++) {
    const node = nodes[i];
    // 如果是string
    if (isString(node)) {
      push(node);
    } else {
      genNode(node, context);
    }
    if (i < nodes.length - 1) push(",");
  }
}

function genNullable(args: any) {
  return args.map((arg) => arg || "null");
}
function genExpression(node: any, context: any) {
  const { push } = context;
  push(`${node.content}`);
}

function genInterpolation(node: any, context: any) {
  const { push, helper } = context;
  push(`${helper(TO_DISPLAY_STRING)}(`);
  genNode(node.content, context);
  push(")");
}

function genText(node: any, context: any) {
  const { push } = context;
  push(`'${node.content}'`);
}
