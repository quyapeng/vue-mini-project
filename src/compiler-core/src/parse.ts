import { NodeTypes } from "./ast";
const openDelimiter = "{{";
const closeDelimiter = "}}";

const enum TagType {
  Start,
  End,
}

export function baseParse(content: string) {
  // 创建一个全局的上下文对象
  const context = createParserContext(content);
  return createRoot(parseChildren(context, []));
}

function parseChildren(context, ancestors) {
  const nodes: any = [];
  // console.log("isend", !isEnd(context, ancestors));
  while (!isEnd(context, ancestors)) {
    let node;
    let s = context.source;
    if (s.startsWith(openDelimiter)) {
      node = parseInterpolation(context);
    } else if (s[0] === "<") {
      // < 开始
      if (/[a-z]/i.test(s[1])) {
        // 第二位是字母，不限制大小
        // console.log("element div", s);
        node = parseElement(context, ancestors);
      }
    }

    //
    if (!node) {
      node = parseText(context);
    }
    nodes.push(node);
  }
  return nodes;
}

function isEnd(context, ancestors) {
  // 1. source 有值
  // 2. 当遇到结束标签的时候
  const s = context.source;
  if (s.startsWith("</")) {
    for (let i = ancestors.length - 1; i >= 0; i--) {
      const tag = ancestors[i].tag;
      if (startsWithEndTagOpen(s, tag)) {
        return true;
      }
    }
  }
  return !s;
}

function parseText(context) {
  //
  let endTokens = ["<", "{{"];
  let endIndex = context.source.length;

  for (let i = 0; i < endTokens.length; i++) {
    // console.log("context.source", context.source);
    let index = context.source.indexOf(endTokens[i]);
    if (index !== -1 && endIndex > index) {
      // 有
      endIndex = index;
    }
  }
  // 1. 获取到值，
  // const content = context.source.slice(0, context.source.length);
  // console.log("parseText", context.source);
  // // 2. 推进， 截取掉
  // advanceBy(context, content.length);
  // console.log("parseText 为空", context.source);
  const content = parseTextData(context, endIndex);
  // console.log("content", content);
  return {
    type: NodeTypes.TEXT,
    content,
  };
}

function parseTag(context, type) {
  // 1.解析 div
  // 2.删除解析过的node
  const match: any = /^<\/?([a-zA-Z]*)/i.exec(context.source);
  const tag = match[1];
  //  match 为 [ '<div', 'div', index: 0, input: '<div></div>', groups: undefined ]
  // 正则匹配出的第一位是'<div',最后一位'>'也需要
  /*
  源码此处是分两步处理 我喜欢一次搞定。
  advanceBy(context, match[0].length);
  advanceBy(context, 1);
  **/
  advanceBy(context, match[0].length + 1);
  // advanceBy(context, match[0].length);
  // advanceBy(context, 1);
  // console.log(context.source);
  if (type == TagType.End) return;
  return {
    type: NodeTypes.ELEMENT,
    tag,
  };
}

function parseTextData(context, length) {
  const content = context.source.slice(0, length);
  // console.log("parseText", context.source);
  // 推进， 截取掉
  advanceBy(context, length);
  return content;
}

function parseElement(context, ancestors) {
  // 1.解析 div
  // 2.删除解析过的node
  const element: any = parseTag(context, TagType.Start);
  ancestors.push(element);
  element.children = parseChildren(context, ancestors);
  ancestors.pop();

  // console.log("element.tag", element.tag);
  // console.log("context.source", context.source);
  // 前后tag一致
  // if (context.source.slice(2, 2 + element.tag.length) === element.tag) {
  if (startsWithEndTagOpen(context.source, element.tag)) {
    parseTag(context, TagType.End);
  } else {
    // 报错
    throw new Error(`缺少结束标签：${element.tag}`);
  }

  // console.log("parseElement", context.source);
  return element;
}

function startsWithEndTagOpen(source, tag) {
  return (
    source.startsWith("</") &&
    source.slice(2, 2 + tag.length).toLowerCase() === tag.toLowerCase()
  );
}

function parseInterpolation(context) {
  // {{message}} 获取到message内容
  // 从index为2开始查找，前面0，1是固定的 {{ ,减少没必要的查询,所以indexof 第二个参数为open的长度
  // 去除前面固定 {{ 的叫法为 推进
  // 最后一个index
  const closeIndex = context.source.indexOf(
    closeDelimiter,
    openDelimiter.length
  );
  //
  //   context.source = context?.source.substr(0, closeIndex);
  // 先删除掉前面的
  //   context.source = context?.source.slice(openDelimiter.length);
  advanceBy(context, openDelimiter.length);
  // 最终解析出来内容的长度
  const rawContentLength = closeIndex - openDelimiter.length;

  // const rawContent = context.source.slice(0, rawContentLength);
  const rawContent = parseTextData(context, rawContentLength);
  const content = rawContent.trim();

  advanceBy(context, closeDelimiter.length);
  //   context.source = context.source.slice(
  //     rawContentLength + closeDelimiter.length
  //   );
  // console.log("context", context);

  return {
    type: NodeTypes.INTERPOLATION,
    content: {
      type: NodeTypes.SIMPLE_EXPRESSION,
      content,
    },
  };
}

function advanceBy(context: any, length: number) {
  context.source = context.source.slice(length);
}
// 创建根结点
function createRoot(children) {
  return {
    children,
    root: NodeTypes.ROOT,
  };
}

function createParserContext(content: string) {
  return {
    source: content,
  };
}
