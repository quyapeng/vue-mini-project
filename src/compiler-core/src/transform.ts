import { NodeTypes } from "./ast";

export function transform(root, { nodeTransforms }) {
  const context = createTransformContext(root, { nodeTransforms });
  // 1. 遍历 - 深度优先搜索
  traverseNode(root, context);
  // 2. 修改text content
}
function traverseNode(node: any, { nodeTransforms = [] }: any) {
  //
  // console.log("node", node);
  // if (node.type == NodeTypes.TEXT) {
  //   node.content = node.content + " mini-vue";
  // }
  for (let i = 0; i < nodeTransforms.length; i++) {
    nodeTransforms[i](node);
  }
  const children = node.children;
  if (children) {
    for (let i = 0; i > children.length; i++) {
      const node = children[i];
      traverseNode(node, { nodeTransforms });
    }
  }
}

function createTransformContext(root, { nodeTransforms = [] }) {
  return {
    root,
    nodeTransforms,
  };
}
