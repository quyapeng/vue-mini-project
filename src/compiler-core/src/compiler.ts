import { generate } from "./codegen";
import { baseParse } from "./parse";
import { transform } from "./transform";
import { transformElement } from "./transforms/transformElement";
import { transformExpression } from "./transforms/transformExpression";
import { transformText } from "./transforms/transformText";

// 统一出口-> transform ,codegen等模块处理
export function baseCompile(template) {
  const ast: any = baseParse(template);
  transform(ast, {
    nodeTransforms: [transformExpression, transformElement, transformText],
  });
  return generate(ast);
}

// 运行模块和编译模块不能互相引用。
