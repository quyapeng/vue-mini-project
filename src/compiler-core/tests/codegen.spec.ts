import { baseParse } from "../src/parse";
import { generate } from "../src/codegen";
import { transform } from "../src/transform";
import { transformExpression } from "../src/transforms/transformExpression";

describe("codegen", () => {
  it("happy path string", () => {
    //
    const ast = baseParse("hi");
    transform(ast);
    const { code } = generate(ast);
    // 快照测试-->
    // 1. 抓问题
    // 2. 有意更新（）
    // toMatchSnapshot 会为expect的结果做一个快照，并与前面的快照做匹配。
    // 执行run之后，生成一个__snapshots__文件夹，里面生成一个codegen.spec.ts.snap文件，这就是快照内容。
    // 执行run的命令之后➕ -u 代表有意更新,此时就会更新快照内容，方面下次对比最新数据。

    expect(code).toMatchSnapshot();
  });

  //
  it("interpolation", () => {
    const ast = baseParse("{{message}}");
    transform(ast, {
      nodeTransforms: [transformExpression],
    });
    const { code } = generate(ast);
    expect(code).toMatchSnapshot();
  });
});
