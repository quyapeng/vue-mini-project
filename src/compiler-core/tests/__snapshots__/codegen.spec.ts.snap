// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen happy path string 1`] = `
"
return function render(_ctx, _cache){return 'hi'}"
`;

exports[`codegen interpolation 1`] = `
"
return function render(_ctx, _cache){return _toDisplayString()}"
`;
